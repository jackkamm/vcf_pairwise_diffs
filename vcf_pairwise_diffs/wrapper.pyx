cimport numpy as np
import numpy as np

from pysam.libchtslib cimport htsFile, hts_open, hts_close, bcf_hdr_t, bcf_hdr_read, bcf_hdr_nsamples, bcf_hdr_destroy

cdef extern from "vcfpairdiffs.h":
    void vcfpairdiffs(htsFile *fp, bcf_hdr_t *hdr,
                      int nsmpl, int* ndiff, int* ntot)

def pairwise_diffs(vcf_name):
    vcf_name = vcf_name.encode()
    cdef const char* fname = vcf_name
    cdef htsFile *fp = hts_open(fname, "r")
    cdef bcf_hdr_t *hdr = bcf_hdr_read(fp)

    nsmpl = bcf_hdr_nsamples(hdr)
    samples = [hdr.samples[i].decode() for i in range(nsmpl)]

    cdef int[:,::1] ndiff = np.zeros([nsmpl, nsmpl], dtype=np.intc)
    cdef int[:,::1] ntot = np.zeros([nsmpl, nsmpl], dtype=np.intc)
    vcfpairdiffs(fp, hdr, nsmpl, &ndiff[0,0], &ntot[0,0])

    bcf_hdr_destroy(hdr)
    hts_close(fp)

    return {"samples": samples,
            "n_differences": np.asarray(ndiff),
            "n_comparisons": np.asarray(ntot)}
