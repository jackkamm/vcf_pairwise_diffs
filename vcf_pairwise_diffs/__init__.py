import numpy as np
import pysam
from .wrapper import pairwise_diffs

# python implementation for prototype/test
def _test_pairwise_diffs(vcf_name):
    bcf_in = pysam.VariantFile(vcf_name)
    samples = bcf_in.header.samples
    n_diffs = np.zeros([len(samples)]*2)
    n_comps = np.zeros([len(samples)]*2)
    for rec in bcf_in.fetch():
        rec_samples = rec.samples
        for i in range(len(samples)):
            for j in range(len(samples)):
                s_i, s_j = rec.samples[i], rec.samples[j]
                idx_i = s_i.allele_indices
                idx_j = s_j.allele_indices
                if idx_i is None or idx_j is None:
                    continue

                for ii in range(len(idx_i)):
                    for jj in range(len(idx_j)):
                        # don't compare to self
                        if i == j and jj == ii:
                            continue
                        a_i = idx_i[ii]
                        a_j = idx_j[jj]
                        if a_i is None or a_j is None:
                            continue
                        n_comps[i,j] += 1
                        if a_i != a_j:
                            n_diffs[i,j] += 1

    return {"samples": list(samples),
            "n_differences": n_diffs,
            "n_comparisons": n_comps}
