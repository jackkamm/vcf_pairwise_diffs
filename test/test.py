import os
import ast
import pprint
import vcf_pairwise_diffs

# test.vcf was copied from htslib/test/test-vcf-api.out
dired = os.path.dirname(os.path.realpath(__file__))
pairwise_diffs = vcf_pairwise_diffs.pairwise_diffs(
    os.path.join(dired, "test.vcf"))
pairwise_diffs2 = vcf_pairwise_diffs._test_pairwise_diffs(
    os.path.join(dired, "test.vcf"))
for k in ('n_comparisons', 'n_differences'):
    pairwise_diffs[k] = pairwise_diffs[k].tolist()
    pairwise_diffs2[k] = pairwise_diffs2[k].tolist()

pprint.pprint(pairwise_diffs)

assert pairwise_diffs == pairwise_diffs2
with open(os.path.join(dired, "test.out")) as f:
    assert ast.literal_eval(f.read()) == pairwise_diffs
