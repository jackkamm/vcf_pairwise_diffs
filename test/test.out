{'n_comparisons': [[6, 15, 12], [15, 6, 12], [12, 12, 6]],
 'n_differences': [[0, 9, 12], [9, 6, 6], [12, 6, 0]],
 'samples': ['NA00001', 'NA00002', 'NA00003']}
